web3/middleware/signing.py

```diff
@@ -0,0 +1,78 @@
+from eth_keys.datatypes import PrivateKey
+from eth_utils import is_same_address
+
+from web3.utils.datastructures import HexBytes
+from web3.exceptions import InvalidAddress
+# from web3.utils.signing import LocalAccount
+
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def construct_transaction_signing_middleware(private_key):
+    """Intercept a transaction in flight and sign it.
+
+    Args:
+        private_key (TYPE): Private key used to sign the transaction.
+                            Accepts the following formats:
+                            - `PrivateKey` object from `eth-keys`
+                            - `LocalAccount` object from `web3.eth.account`
+                            - raw bytes
+                            - TODO: hex encoded string
+
+
+    Returns:
+        TYPE: Description
+    """
+
+    # Check if the private_key is either
+    # - eth_keys.keys.PrivateKey
+    # - web3.eth.account.create().privateKey -- can just check that its HexBytes
+    # - bytes
+    # TODO: - hexbytes
+    valid_key_types = (PrivateKey, HexBytes, bytes)
+
+    _private_key = private_key if isinstance(private_key, valid_key_types) else None
+    if not _private_key:
+        raise ValueError('Private Key is invalid.')
+
+    def transaction_signing_middleware(make_request, web3):
+        def middleware(method, params):
+            # Only operate on the `eth.sendTransaction` method
+            # Only operate on if the private key matches the public key in the transaction
+            # Note: params == transaction in this case.
+            if method != 'eth_sendTransaction':
+                return make_request(method, params)
+
+            transaction = params
+            transaction_from_address = transaction.get('from')
+            private_key_address = web3.eth.account.privateKeyToAccount(_private_key).address
+
+            if not is_same_address(transaction_from_address, private_key_address):
+                return make_request(method, params)
+
+            if 'gas' not in transaction:
+                try:
+                    transaction['gas'] = web3.eth.estimateGas(transaction)
+                except ValueError:
+                    # Raise Error?
```

Comment:
This error should probably bubble.  If no gas was provided and `estimateGas` fails then we have to fail hard.

Reply:
Yea, the `estimateGas` method does bomb in my test suite.

`transaction['gas'] = web3.eth.estimateGas(transaction)` fails with a long traceback.  The end of it reads:

```
E           ValueError: b'\x1d\xa8\xbd\xea\xe6\xa5\xf5.\xa2\xca\x1fR\x0c\xc9i\xa5\xb0\xe3\x14M' is not in list

_from      = b'\x1d\xa8\xbd\xea\xe6\xa5\xf5.\xa2\xca\x1fR\x0c\xc9i\xa5\xb0\xe3\x14M'
data       = b''
gas        = None
gas_price  = None
self       = <testrpc.client.client.EthTesterClient object at 0x7f0eb873a208>
to         = b'0x82A978B3f5962A5b0957d9ee9eEf472EE55B42F1'
value      = 123
```

Looks like there is a formatting different between the `from` and the `to` fields.  Could that be the issue?